app/controllers/invoices_controller.rb

```diff
@@ -1,7 +1,73 @@
 class InvoicesController < ApplicationController
+  before_action :set_invoice, only: %i[show edit update]
+  before_action :set_vehicle
+
+  def index
+    @invoices = Invoice.where(vehicle_id: params[:vehicle_id])
+  end
+
+  def show
+    @shop = Shop.find(@invoice.shop_id)
+  end
+
+  def new
+    @invoice = Invoice.new
+    @action = 'Ajouter'
+  end
+
+  def create
+    @invoice = Invoice.new(invoice_params)
+    @invoice.vehicle = @vehicle
+    @vehicle.total_expenses += params[:invoice][:price].to_f
+
+    if @invoice.save
+      redirect_to vehicle_invoice_path(@vehicle, @invoice), notice: 'Facture créee'
```

Comment:
_total_expenses_ calculation is not at the right place. If the invoice is invalid, you will call the method **create** twice (or more), and the total will be increased each time the method is called.
Please move line 21 inside the _IF_ in order to update the total only when the invoice is valid ans saved.

Same change to do in the method **update**, move the call to _old_invoice_price_ inside the _IF_.